src/main/java/net/wurstclient/altmanager/MicrosoftLoginManager.java

```diff
@@ -0,0 +1,335 @@
+        postData.put("scope", scopeURL);
+
+        String encodedData = encodeMap(postData);
+
+        try {
+            byte[] encodedDataBytes = encodedData.getBytes(StandardCharsets.UTF_8);
+            URL url = new URL("https://login.live.com/oauth20_token.srf");
+            HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
+            httpURLConnection.setRequestMethod("POST");
+            httpURLConnection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8");
+            httpURLConnection.setDoOutput(true);
+
+            try(OutputStream os = httpURLConnection.getOutputStream()) {
+                os.write(encodedDataBytes);
+            }
+
+            String data = new BufferedReader(new InputStreamReader(httpURLConnection.getInputStream())).lines().collect(Collectors.joining());
+
+            JsonParser jsonParser = new JsonParser();
+            JsonElement jsonElement = jsonParser.parse(data);
+            JsonObject object = jsonElement.getAsJsonObject();
+
+            authToken = object.get("access_token").getAsString();
+        }catch(IOException e) {
+            e.printStackTrace();
+        }
+        return authToken;
+    }
+
+    private static String getAuthorizationCode(String email, String password) {
+        String cookie = "";
+        String ppft = "";
+        String login = "";
+
+        try {
+            URL url = new URL(loginURL);
+            HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
+            InputStream inputStream = httpURLConnection.getInputStream();
+
+            cookie = httpURLConnection.getHeaderField("set-cookie");
+
+            String data = new BufferedReader(new InputStreamReader(inputStream)).lines().collect(Collectors.joining());
+            Matcher matcher = Pattern.compile("sFTTag:[ ]?'.*value=\"(.*)\"/>").matcher(data);
+            if(matcher.find()) {
+                ppft = matcher.group(1);
+            }else{
+                throw new AuthenticationException("sFTTag not found in response.");
+            }
+
+            matcher = Pattern.compile("urlPost:[ ]?'(.+?(?='))").matcher(data);
+            if(matcher.find()) {
+                login = matcher.group(1);
+            }else{
+                throw new AuthenticationException("urlPost not found in response.");
+            }
+        } catch (IOException | AuthenticationException e) {
+            e.printStackTrace();
+        }
+        return microsoftLogin(email, password, cookie, ppft, login);
+    }
+
+    private static String microsoftLogin(String email, String password, String cookie, String ppft, String loginURL) {
+        String authCode = "";
+
+        Map<String, String> postData = new HashMap<>();
+        postData.put("login", email);
+        postData.put("loginfmt", email);
+        postData.put("passwd", password);
+        postData.put("PPFT", ppft);

```

Comment:
Pardon my ignorance, but what does PPFT / sFTTag stand for and what does it do? I can't seem to find any documentation for this stuff.

Reply:
This part is to actually login to Microsoft, otherwise you have to load it in a web browser

Reply:
I don't know exactly what it stands for, however examples such as [this](https://github.com/amen/Autoclaimer/blob/d50bcdf607b14e64828bac9c70997d79ab885c37/src/main/java/XboxLiveAuth.java) and [this](https://github.com/Badbird-5907/MinecraftAuthLib/blob/32dd9d60d10671b74bc44102c77fb27553c48eac/src/main/java/net/badbird5907/authlib/requestors/UserAuthRequestor.java) should hopefully explain it

Reply:
[https://wiki.vg/Zh:Microsoft_Authentication_Scheme](This) also explains it well, it has more information than the English version (translated version [here](https://wiki-vg.translate.goog/Zh:Microsoft_Authentication_Scheme?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=en&_x_tr_pto=wapp)